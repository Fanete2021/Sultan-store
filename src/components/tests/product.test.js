/* eslint-disable testing-library/render-result-naming-convention */
const React = require('react')
const { render, screen, fireEvent } = require('@testing-library/react')
const Product = require('../Product').default
const { BrowserRouter } = require('react-router-dom')
const reduxHooks = require('react-redux')
const { cartSlice } = require('../../store/reducers/slices/CartSlice')
const { editorSlice } = require('../../store/reducers/slices/EditorSlice')
const { productState } = require('./states')

jest.mock('react-redux')

const mockedDispatch = jest.spyOn(reduxHooks, 'useDispatch')

describe('editor', () => {
    it('should create Product', () => {
        mockedDispatch.mockResolvedValue(jest.fn())

        const element = render(
            <BrowserRouter>
                <Product product={productState} />
            </BrowserRouter>
        )

        expect(element).toMatchSnapshot()
    })

    it('should dispatch actions', () => {
        const dispatch = jest.fn()
        mockedDispatch.mockReturnValue(dispatch)

        const mockedAddComplete = jest.spyOn(cartSlice.actions, 'addItem')

        const mockedSetEditableProductComplete = jest.spyOn(
            editorSlice.actions,
            'setEditableProduct'
        )
        const mockedChangeShowingEditorComplete = jest.spyOn(editorSlice.actions, 'changeShowing')

        render(
            <BrowserRouter>
                <Product product={productState} />
            </BrowserRouter>
        )

        fireEvent.click(screen.getByTestId('cart'))
        fireEvent.click(screen.getByTestId('editor'))

        expect(dispatch).toHaveBeenCalledTimes(3)

        const cartItem = {
            product: productState,
            count: 1,
        }

        expect(mockedAddComplete).toHaveBeenCalledWith(cartItem)
        expect(mockedSetEditableProductComplete).toHaveBeenCalledWith(productState)
        expect(mockedChangeShowingEditorComplete).toHaveBeenCalledWith(true)
    })
})
