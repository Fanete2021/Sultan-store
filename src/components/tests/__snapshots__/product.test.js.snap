// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`editor should create Product 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="product"
      >
        <div
          class="product__image"
        >
          <a
            href="/catalog/1"
          >
            <img
              alt=""
              src=""
            />
          </a>
        </div>
        <div
          class="product__wrapper"
        >
          <div
            class="product__size"
          >
            <img
              alt=""
              src=""
            />
            <div
              class="size__value"
            >
              22
               
              г
            </div>
          </div>
          <div
            class="product__name"
          >
            <span
              class="name__title"
            >
              <b>
                AOS
              </b>
            </span>
             
            AOS
          </div>
          <div
            class="product__barcode"
          >
            <span
              class="description"
            >
              Штрихкод:
            </span>
             1231314
          </div>
          <div
            class="product__manufacturer"
          >
            <span
              class="description"
            >
              Производитель:
            </span>
             AOS
          </div>
          <div
            class="product__brand"
          >
            <span
              class="description"
            >
              Бренд:
            </span>
             AOS
          </div>
          <div
            class="product__actions"
          >
            <div
              class="actions__price"
            >
              100
               ₸
            </div>
            <div
              class="actions__button"
              data-testid="cart"
            >
              <div
                class="button__description"
              >
                В корзину
              </div>
              <img
                alt=""
                src=""
              />
            </div>
            <img
              alt=""
              class="actions__edit"
              data-testid="editor"
              src=""
            />
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="product"
    >
      <div
        class="product__image"
      >
        <a
          href="/catalog/1"
        >
          <img
            alt=""
            src=""
          />
        </a>
      </div>
      <div
        class="product__wrapper"
      >
        <div
          class="product__size"
        >
          <img
            alt=""
            src=""
          />
          <div
            class="size__value"
          >
            22
             
            г
          </div>
        </div>
        <div
          class="product__name"
        >
          <span
            class="name__title"
          >
            <b>
              AOS
            </b>
          </span>
           
          AOS
        </div>
        <div
          class="product__barcode"
        >
          <span
            class="description"
          >
            Штрихкод:
          </span>
           1231314
        </div>
        <div
          class="product__manufacturer"
        >
          <span
            class="description"
          >
            Производитель:
          </span>
           AOS
        </div>
        <div
          class="product__brand"
        >
          <span
            class="description"
          >
            Бренд:
          </span>
           AOS
        </div>
        <div
          class="product__actions"
        >
          <div
            class="actions__price"
          >
            100
             ₸
          </div>
          <div
            class="actions__button"
            data-testid="cart"
          >
            <div
              class="button__description"
            >
              В корзину
            </div>
            <img
              alt=""
              src=""
            />
          </div>
          <img
            alt=""
            class="actions__edit"
            data-testid="editor"
            src=""
          />
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
